kvclient: dont' evict down leaseholder

andrewbaptist opened the pull request:
We should not change routing information based on whether a leaseholder is up or down. This causes churn in the updating of lease information. Instead we reply on circuit breakers to fast-ignore down leaseholders

Epic: none

Release note: None

Commit:
kvclient: don't evict down leaseholder

We should not change routing information based on whether a leaseholder
is up or down. This causes churn in the updating of lease information.
Instead we reply on circuit breakers to fast-ignore down leaseholders

This removal of the leaseholder was addressed a few years ago with the
addition of `ClientRangeInfo` and the caching of the leaseholder in the
range cache.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/118828)
<!-- Reviewable:end -->

andrewbaptist commented:
I'm going to run a little more testing on this. But we need to have this behavior change to avoid churning the leaseholder for partial partitions. I don't think this was actually necessary anymore since we automatically "skip" down leaseholders in the transport anyway.

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118828)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2547 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118828#-NpzEXRD68Kk9IlPQNN1:-NpzEXRE5f7LpNet1lA6:b-2lwhks) ([raw file](https://github.com/cockroachdb/cockroach/blob/b876c990bbd40af075dda6ab6b28b392ec6da19f/pkg/kv/kvclient/kvcoord/dist_sender.go#L2547)):*
> ```Go
> 			// replica into the cache, but without a leaseholder (and taking into
> 			// account that the local node can't be down) it won't take long until we
> 			// talk to a replica that tells us who the leaseholder is.
> ```

Is the above case still relevant? If it is, what's the plan for addressing it?


<!-- Sent from Reviewable.io -->

andrewbaptist commented:
This appears to be fine (passes all tests and works on a real cluster). Let me know if there is any additional testing we should add arounds this before we merge it. This is required for partial partition work (since we don't want to keep evicting and re-adding the leaseholder) - but maybe there are alternate solutions if this isn't OK.

erikgrinaker commented:
Do we have a test case for the scenario described in the comment? How is it handled without this logic?

erikgrinaker reviewed (approved):
Nevermind, this case is now handled by `ClientRangeInfo`. If we have a stale cached leaseholder pointing to an unavailable n1, but the lease has since moved to n2, the successful response from n2 will update our cache with the new lease via `ClientRangeInfo`. This code predates `ClientRangeInfo` by 2 years.

andrewbaptist commented:
TFTR!

The one test that changes covers the behavior now. Specifically we don't evict from the cache in `TestDistSenderDownNodeEvictLeaseholder` anymore. I renamed the test to `TestDistSenderLeaseholderDown` to make this clear and updated the comment.

Let me know if you think I should add another test for this.

andrewbaptist commented:
bors r=erikgrinaker

TFTR!
